### DISCUSSION TITLE
Handling storage exceptions in Keycloak

### DISCUSSION THREAD

**Discussion by ahus1** [keycloak/keycloak#9601] [Archive]
# Problem statement

When accessing data from Keycloak (either read or write), there can be different kinds of problems, usually in the kind of transactions failing, database constraints violated, lazy loading issues. 

In the example of the JPA map storage when listing all the clients for a REST Admin API caller, it first reads a subset of the fields for performance reasons. Later, when retrieving all details, it will fetch the additional fields. In this case the record in the table could have been deleted or modified.

It could be seen when running ConcurrencyTest#createRemoveClient() with a JPA map storage configuration: The server returns a 500 error when retrieving a list of clients while one of the clients in the list is removed by another client concurrently.

# Approach

When the JPA map storage provider recognizes that the entity has been removed from the database, it should throw an exception that extends the ModelException (for example: ObjectRemovedException). The ClientResource#getClients() could then catch the exception and filter the instance from the list. 

There seems to be an existing handling of exceptions in place, although it is 
* quite unspecific in the types of exceptions (see: `try { c.getClientId(); return true; } catch (Exception ex) { return false; }`)
* very specific in the field it tests - getClientId() tests only lazy loading for the traditional JPA implementation.

In the mid-term there might be additional exceptions like ObjectModifiedException for optimistic locking exceptions. In that case the ClientResource will be more difficult to recover and would most likely pass a 500 error to the caller.

# Benefits

The caller would retrieve a list of valid clients even if one client of the list is removed concurrently.

# Alternatives

The caller could receive a 500 exception. The caller would then still additional information if the problem is a transient problem and when to retry the call.

**Comment by hmlnarik** [comment]
Thank you for the idea. The same applies to the use case of a user obtained from the cache but in the meantime removed externally (e.g. from LDAP) - once the fields in the LDAP object would be accessed that are not stored in this cache, it would be the same situation of a partially loaded object and the same measures should apply.

Hence I agree that the case of partially loaded model should be clearly signalled to the caller.

Before going forward, it is important to understand where the raising and handling the exception takes place. From that we will get the right exception parent (whether it should be a `ModelException`), and also open up discussion for potential remedy measures.

Could you please add the following information for a client use case (mentioned in the description) and user use case (mentioned here):
- Where would this exception be thrown from?
- Where would this exception be caught?
- How would the exception handling apply to the current code (which layers - services, logical, physical - would be affected)?

**Reply by ahus1** [reply]
> Where would this exception be thrown from?

The exception in the example above will he thrown in `JpaClientDelegateProvider` when checking the result of `delegate = em.find(JpaClientEntity.class, UUID.fromString(delegate.getId()));`. If the entity is gone, em.find() will return null. 

Similar exceptions could be raised in other places throughout a storage provider implementation; it might also make sense to wrap some of the Hibernate/JPA runtime exceptions later in case we want to handle them in a higher level of the stack.

> Where would this exception be caught?
> How would the exception handling apply to the current code (which layers - services, logical, physical - would be affected)?

I assume catching and handling would be done in the services as handling would require context, but I am not sure if I fully understand the layers correctly yet. I would target at code handling lists; if this happens when handling a single entity I would rather throw the exception and show an error to the caller.

Handling of the exception example above will be done in ClientResource. The `try {}` would be placed around the code in map(), and would return null on any removed entity, that would later be filtered out. The code could also be changed to flatMap() to return either an empty or one element stream to avoid the filtering of nulls, but this doesn't change the place where it would be handled.

```
        // existing code
        Stream<ClientRepresentation> s = clientModels
                /* ... */
                .map(c -> {
                    ClientRepresentation representation = null;
                    if (canView || auth.clients().canView(c)) {
                        representation = ModelToRepresentation.toRepresentation(c, session);
                        representation.setAccess(auth.clients().getAccess(c));
                    } else if (!viewableOnly && auth.clients().canView(c)) {
                        representation = new ClientRepresentation();
                        representation.setId(c.getId());
                        representation.setClientId(c.getClientId());
                        representation.setDescription(c.getDescription());
                    }

                    return representation;
                })
                .filter(Objects::nonNull);
```

**Reply by ahus1** [reply]
(I thought about the service/logical/physical layer a bit more: Usually I would go by package name or maven module to identify what is in a  layer. With services it is straight-forward to me: everything in a package with "service" in it would be a service. Everything with "jpa" in its package name sounds quite physical, but would be "storage" according to https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/storage-architecture.md? I wonder what classes in the "org.keycloak.models.map.*" packages would be. It would be simpler for me if the layers would repeat in the package names as they do for "service")

**Reply by hmlnarik** [reply]
Thank you.

> The exception in the example above will he thrown in JpaClientDelegateProvider when checking the result of delegate = em.find(JpaClientEntity.class, UUID.fromString(delegate.getId()));. If the entity is gone, em.find() will return null.

`JpaClientDelegateProvider` operates on physical layer where there is no session or models, which would justify using `ModelException` as the base. On the other hand, database exceptions are already [turned into `ModelException` in the existing provider](https://github.com/keycloak/keycloak/blob/f80d336276fb73603415aedda56b6da4580deb00/model/jpa/src/main/java/org/keycloak/connections/jpa/PersistenceExceptionConverter.java). Hence my suggestion would be to introduce a `ModelException` child, called a `ModelIllegalStateException` in [`org.keycloak.models`](https://github.com/keycloak/keycloak/tree/main/server-spi/src/main/java/org/keycloak/models) package which would serve as a base for all exceptions that involve an invalid state of a model, and that includes illegal internal state. This exception would be then extended in the model-map module into a `StorageException` (or perhaps `TransientStorageException` to follow your suggestion on separating the transient and persistent error states). This way we can eventually split the two if that becomes desirable.

> I assume catching and handling would be done in the services [...]  I would target at code handling lists; if this happens when handling a single entity I would rather throw the exception and show an error to the caller. [...]

The layers are assigned correctly. Here comes the question though: What would be the proposal to not involve including `try - catch` block in the `map` and `filter(Objects::nonNull)` directly? The goal here is to get rid of the boilerplate code. One option could be to do something like (note the `illegalModelToNull` method) but that is only half-way solution:
```java
        Stream<ClientRepresentation> s = clientModels
                /* ... */
                .map(illegalModelToNull(c -> {
                    ClientRepresentation representation = null;
                    if (canView || auth.clients().canView(c)) {
                        representation = ModelToRepresentation.toRepresentation(c, session);
                        representation.setAccess(auth.clients().getAccess(c));
                    } else if (!viewableOnly && auth.clients().canView(c)) {
                        representation = new ClientRepresentation();
                        representation.setId(c.getId());
                        representation.setClientId(c.getClientId());
                        representation.setDescription(c.getDescription());
                    }

                    return representation;
                }))
                .filter(Objects::nonNull);
```
Is there any better solution?

**Reply by ahus1** [reply]
I think the key to a cleaner processing is using `flatMap` instead of `map`, as it allows to remove elements of the stream by returning an empty stream instead of null. 

The following assumes that also `.filter(c -> { try { c.getClientId(); return true; } catch (Exception ex) { return false; } } )` can be rolled up into this by mapping the lazy loading exception to the new ModelException in `PersistenceExceptionConverter` as well. 

This would result into code like this: 

```
        Stream<ClientRepresentation> s = clientModels
                .flatMap(c -> handleModelExceptions(() -> {
                    ClientRepresentation representation = null;
                    if (canView || auth.clients().canView(c)) {
                        representation = ModelToRepresentation.toRepresentation(c, session);
                        representation.setAccess(auth.clients().getAccess(c));
                    } else if (!viewableOnly && auth.clients().canView(c)) {
                        representation = new ClientRepresentation();
                        representation.setId(c.getId());
                        representation.setClientId(c.getClientId());
                        representation.setDescription(c.getDescription());
                    }
                    return Optional.ofNullable(representation);
                }));
```

which a helper function 

```
    /**
     * Attempts to handle exceptions that occur when transforming the model to a representation.
     *
     * Entities that have been removed from the store will not show up in the output stream.
     */
    public static <R> Stream<R> handleModelExceptions(Supplier<Optional<R>> modelToRepresentation) {
        try {
            Optional<? extends R> representation = modelToRepresentation.get();
            return representation.<Stream<R>>map(Stream::of).orElseGet(Stream::empty);
        } catch (ModelException ex) { // TODO: replace with exception that indicates that element has been removed
            return Stream.empty();
        }
    }
``` 

The helper function could be placed in ModelToRepresentation, as it will be used together with the toRepresentation() methods that are present there as well.

Another thought on the transient error indicator: The concept of exceptions being transient or permanent might be orthogonal to the exception hierarchy that is evolving, that is: there could be different exceptions all over the ModelException hierarchy that are transient, not necessarily in the storage area. It could evolve as a flag in ModelException or as a marker interface.

Therefore: let's focus on picking one exception class that clearly identifies the situation "entity has been removed" and allows to return the empty stream in the handleModelExceptions(), while any other exception related to storage (for example entity has been modified) would not be caught here and would bubble up and lead to a 500 error for the caller.

**Comment by hmlnarik** [comment]
The `flatMap` is a good idea. It means creation of many short-lived stream objects though - one per each model. In this light, it seems that mandatorily using the `filter` actually pays off better since there is no additional overhead of instantiation of small objects.

What is the value of introducing the `Optional` in the helper function?

I like the idea of extending `ModelToRepresentation` with appropriate function, in this case it would be
```java
<M, R> Stream<R> ModelToRepresentation.toRepresentation(Stream<M> models, Function<M, R> transformer) {
    return models.map(m -> try { return transformer(m);} catch (ModelIllegalStateException e) { return null; })
         .filter(Objects::nonNull);
}
```

> Therefore: let's focus on picking one exception class that clearly identifies the situation "entity has been removed" and allows to return the empty stream in the handleModelExceptions(), while any other exception related to storage (for example entity has been modified) would not be caught here and would bubble up and lead to a 500 error for the caller.

This need not be the removal, but for certain the entity supporting the model became illegal. The entity may become illegal by expiring or other ways, so we should not concentrate only on a single painpoint of now. That's why `ModelIllegalStateException` represents better the 
state from the perspective of services. Indeed, then there might be e.g. `EntityNotFoundException` declared in map storage and thrown from the low-level code if there is clear that the entity has not been found. This still leaves room for the rest.

As for the transient error indicator, I agree we can postpone this discussion for now.

**Comment by ahus1** [comment]
Introducing the Optional makes the code more readable IMHO; I like to avoid returning nulls. I wanted to make it explicit that the transformation returns one instance or no instance. 

Whether the short-lived streams are a performance penalty or not would be hard for me to tell without a JMH test. At the same time the function you propose is elegant and readable, and with some JavaDoc to indicate that the transformation can return null it is good to go for me. 

To better understand the `ModelIllegalStateException` - "illegal state" has some nuances and vagueness for me, therefore I ask: Given an optimistic lock exception that would occur then the object has been modified in the meantime but still exists: Would throw a different exception or an illegal state exception? Could you think of two other clarifying examples what would be such an illegal state, and what wouldn't?

**Reply by hmlnarik** [reply]
> Introducing the Optional makes the code more readable IMHO; I like to avoid returning nulls. 

Here we are on the same page and I'm happy to bring it in for justified use cases. Here `null`  helps in avoiding creation of small short-lived objects while making NPE impossible, so I find it safe and more performant than the variant with `Optional`.

> Given an optimistic lock exception that would occur then the object has been modified in the meantime but still exists: Would throw a different exception or an illegal state exception?

The `ModelIllegalStateException` would be appropriate.

> Could you think of two other clarifying examples what would be such an illegal state, and what wouldn't?

Illegal state of a model is removal of the entity in the meantime (as being discussed), or expiration of loaded object by some timer if entity validity is time-bound. It is illegal state of the object from the perspective of the caller.

An attempt to insert a duplicate record would not be `ModelIllegalStateException` as the model (as is) would be still valid.

**Reply by ahus1** [reply]
If the `ModellIllegalStateException` would lead to filtering out entities that have been modified concurrently, this would be confusing for the clients: they ask for a list of clients, they receive a list of clients, but they can't be sure if the list is complete as some clients might have been removed due to a transient concurrent update problem. 

Therefore I suggest to remove clients from the list only in those cases where the client has been removed from the store. 

Given the example of a time-bound entity, it would be ok to remove it from the list, as this would be consistent with the expectations of the caller.

**Reply by hmlnarik** [reply]
This is interesting. The caller would not be able to retrieve the details from the entity for a particular model (because of the changed version). Do you suggest to return a partial state (whatever that means)?

Is this any more confusing than in case when there is another client created concurrently while listing the clients in another thread? They would receive the list of valid clients without that client being included.

**Reply by ahus1** [reply]
For a concurrent modification, a store might try to reload the entity in a consistent way. With the map example at hand and JPA this could be impossible as part of the model has already been mapped, and the exception occurs when retrieving additional parts of the model. Throwing an exception and showing a 500 to the caller would be the only option I see here. 

Even if a storage might be able to reload the data consistently on a per-entity level, the entity might have been modified in a way that it would no longer fit the search criteria used in the search. 

Therefore I would rule out recovery from a concurrent modification situation as it would require reading the state as it what when the search was performed. 

I would opt for an exception and a 500 response code in cases of concurrent modification.

**Reply by ahus1** [reply]
> Is this any more confusing than in case when there is another client created concurrently while listing the clients in another thread?

With a new client being added concurrently, the returned list would be a consistent state that was valid at the time of querying. There would also be some causal consistency: after the client has been added, the client can be retrieved via the list of clients until it is removed. 

If a concurrent modification would remove a client from the list of returned clients, this would no longer hold true.

**Reply by hmlnarik** [reply]
Now I see better how you would like it to resolve, while it is a good idea, it is not viable.

The transaction is only started and finished with the start / end of request processing (see [Keycloak request filter](https://github.com/keycloak/keycloak/blob/8ed7c0544fb10e66e1e033b2c0f3b0fbdeeb6e28/services/src/main/java/org/keycloak/services/filters/AbstractRequestFilter.java#L32-L52)). Reloading the entity is thus not an option, as you have stated as well, since reloading would require a retry, and this is not possible at this moment.

It is also not nice to throw 500 since part of the response including headers might have already been sent to the client, and also this could  increase a DoS attack surface (if you modify clients fast enough, there would be decent chance to block valid requests to list the clients). For the same reason, I'd vouch against retrying the request in such a case.

Hence the remaining option is to do the best effort which is finish sending the stream successfully, with all the records most of the time. The UI / client can show a warning if the number of returned clients does not match the expected one.

**Reply by ahus1** [reply]
> it is not viable [...]

Thanks for the the reasoning for that decision. I'll see that it ends up at least as a piece of JavaDoc next to the exception and the function we agreed to put into `ModelToRepresentation`: to explain how the exception should be used, and the reasons behind it. 

> [...] with all the records most of the time. 

If this is the path to go, this should be documented somewhere for the caller as it would come as a surprise. The current documentation of the Admin API is generated from the JavaDoc of the ClientsResource. Would this be the place to add it?

> The UI / client can show a warning if the number of returned clients does not match the expected one

AFAIK the admin API's response doesn't provide that information, just a stream of ClientRepresentations, therefore the caller will have no way to know that a client got missing.

**Reply by ahus1** [reply]
After talking to @hmlnarik today, we agreed to document the decisions taken in JavaDoc. While some of the JavaDoc documentation from ClientResource ends up in the REST API documentation automatically, additional documentation is out of scope for this issue. 

The situation that the caller will not know about filtered out rows needs to be addressed in a future redesign of the REST API of Keycloak. A JavaDoc note in ClientResource should guide that future implementation.

Next step: I'll create a GitHub issue summarizing the tasks for the implementation.

**Comment by ahus1** [comment]
GitHub issue created and assigned to me: https://github.com/keycloak/keycloak/issues/9645
